File: intellij-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/CliLog.kt
package com.ezevillo.gitreview.domain

const val STDERR_MAX: Int = 2000

/** Quote one argv for log readability (not shell-reparseable). */
fun shellQuoteArg(arg: String): String {
    if (arg.isEmpty()) return "\"\""
    if (Regex("""[\s"\\]""").containsMatchIn(arg)) {
        return "\"${arg.replace("\\", "\\\\").replace("\"", "\\\"")}\""
    }
    return arg
}

fun formatCommandLine(command: String, args: List<String>): String =
    (listOf(command) + args.map { shellQuoteArg(it) }).joinToString(" ")

data class CliLogEnd(
    val exitCode: Int?,
    val errorCode: String? = null,
    val durationMs: Long,
    val stderr: String,
    val timedOut: Boolean = false,
)

/** Formats end-of-invocation log lines (pure; host chooses the sink). */
fun formatCliEnd(result: CliLogEnd): List<String> {
    val ms = "${result.durationMs}ms"
    if (result.timedOut) {
        return listOf("← timed out after $ms (killed)")
    }
    if (result.errorCode != null) {
        return listOf("← spawn failed ${result.errorCode}  $ms")
    }
    val line = "← exit ${result.exitCode ?: "null"}  $ms"
    if (result.exitCode == 0) {
        return listOf(line)
    }
    val lines = ArrayList<String>()
    lines.add(line)
    val err = result.stderr.trimEnd()
    if (err.isEmpty()) return lines
    val body = if (err.length > STDERR_MAX) {
        "${err.take(STDERR_MAX)}\n… (truncated)"
    } else err
    for (part in body.split(Regex("\r?\n"))) {
        lines.add("  $part")
    }
    return lines
}

File: visualstudio-extension/src/GitReview.Domain/CliLog.cs
using System.Text.RegularExpressions;

namespace GitReview.Domain;

public static class CliLog
{
    public const int StderrMax = 2000;

    public static string ShellQuoteArg(string arg)
    {
        if (arg.Length == 0) return "\"\"";
        if (Regex.IsMatch(arg, @"[\s""\\]"))
            return "\"" + arg.Replace("\\", "\\\\").Replace("\"", "\\\"") + "\"";
        return arg;
    }

    public static string FormatCommandLine(string command, IReadOnlyList<string> args) =>
        string.Join(" ", new[] { command }.Concat(args.Select(ShellQuoteArg)));

    public sealed record CliLogEnd(
        int? ExitCode,
        string? ErrorCode,
        long DurationMs,
        string Stderr,
        bool TimedOut = false);

    public static IReadOnlyList<string> FormatCliEnd(CliLogEnd result)
    {
        var ms = $"{result.DurationMs}ms";
        if (result.TimedOut)
            return new[] { $"← timed out after {ms} (killed)" };
        if (result.ErrorCode is not null)
            return new[] { $"← spawn failed {result.ErrorCode}  {ms}" };
        var line = $"← exit {result.ExitCode?.ToString() ?? "null"}  {ms}";
        if (result.ExitCode == 0)
            return new[] { line };
        var lines = new List<string> { line };
        var err = result.Stderr.TrimEnd();
        if (err.Length == 0) return lines;
        var body = err.Length > StderrMax
            ? err[..StderrMax] + "\n… (truncated)"
            : err;
        foreach (var part in body.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None))
            lines.Add($"  {part}");
        return lines;
    }
}
